tutorials/editor_basics/editor_events.md

```diff
@@ -2,172 +2,168 @@
 title: 3. Handling Editor Events
 ---
 
-The following set of tutorials is meant to be an introduction to actions activated by editor events.
-The IntelliJ Platform SDK provides a set of callbacks for handling events related to the Editor.
+The previous tutorial [Editor Coordinate Systems](coordinates_system.md) described working with caret coordinate systems in an editor window.
+Caret position was discussed in terms of Logical Position, Visual Position, and Offset.
 
-[Source code](https://github.com/JetBrains/intellij-sdk-docs/tree/master/code_samples/editor_basics)
+This tutorial introduces the Editor Action system, which handles actions activated by keystroke events in the editor.
+The development of two classes from the [editor_basics](https://github.com/JetBrains/intellij-sdk-docs/tree/master/code_samples/editor_basics) code sample are used to illustrate:
+* Using an IntelliJ Platform [EditorActionHandler](upsource:///platform/platform-api/src/com/intellij/openapi/editor/actionSystem/EditorActionHandler.java) to manipulate a caret.
+* Creating and registering a custom [TypedActionHandler](upsource:///platform/platform-api/src/com/intellij/openapi/editor/actionSystem/TypedActionHandler.java) to intercept keystrokes and change the document.
 
+The tutorial is presented in the following sections:
+* bullet list
+{:toc}
 
-## 3.1. Handling keystrokes in the Editor
+## Using an IntelliJ Platform EditorActionHandler
+In this portion of the tutorial the [editor_basics](https://github.com/JetBrains/intellij-sdk-docs/tree/master/code_samples/editor_basics) code sample is used to demonstrate cloning an existing caret.
+A custom action class will use `EditorActionManager` to access a specific `EditorActionHandler` for caret cloning.
+The `editor_basics` code sample adds an **Editor Add Caret** menu item to the editor context menu:
 
-[TypedActionHandler](upsource:///platform/platform-api/src/com/intellij/openapi/editor/actionSystem/TypedActionHandler.java)
-interface is intended to implement custom handling for keys typed in the editor.
-The steps below show how to customize the behaviour of the editor and make it react to typing in a different way than just displaying the typed character in the editor area.
-
-### 3.1.2 Implementing *TypedActionHandler*
-
-First we need to implement an instance of
-[TypedActionHandler](upsource:///platform/platform-api/src/com/intellij/openapi/editor/actionSystem/TypedActionHandler.java):
+![Editor Basics Menu](img/editor_add_caret.png){:width="500px"}
```

Comment:
remove?

Reply:
Replaced with one image for all 3 pages.